opte/src/engine/icmpv6.rs

```diff
@@ -346,18 +349,141 @@ impl HairpinAction for RouterAdvertisement {
+    let icmp =
+        Icmpv6Repr::parse(&smol_src, &smol_dst, &src_pkt, &Csum::ignored())?;
+
+    let (target_addr, has_ll_option) = match icmp {
+        Icmpv6Repr::Ndisc(NdiscRepr::NeighborSolicit {
+            lladdr,
+            target_addr,
+        }) => (Ipv6Addr::from(target_addr), lladdr.is_some()),
+        _ => {
+            // We should never hit this case because the predicate
+            // should have verified that we are dealing with a
+            // Neighbor Solicitation. However, programming error could
+            // cause this to happen -- let's not take any chances.
+            return Err(GenErr::Unexpected(format!(
+                "expected a NDP Neighbor Solicitation, got {} {}",
+                src_pkt.msg_type(),
+                src_pkt.msg_code()
+            )));
+        }
+    };
+
+    // The target cannot be a multicast address.
+    if target_addr.is_multicast() {
+        return Err(GenErr::Unexpected(String::from(
+            "Received NS with multicast target address, dropping.",
+        )));
+    }
+
+    // NS is only allowed from the unspecified address if the destination is a
+    // solicited-node multicast address.
+    if metadata.src == Ipv6Addr::ANY_ADDR
+        && !metadata.dst.is_solicited_node_multicast()
+    {
+        return Err(GenErr::Unexpected(String::from(
+            "Received NS from UNSPEC, but destination is not the solicited \
+            node multicast address, dropping.",
+        )));
+    }
+
+    // Cannot contain Link-Layer address option if from the unspecified address.
+    if metadata.src == Ipv6Addr::ANY_ADDR && has_ll_option {
+        return Err(GenErr::Unexpected(String::from(
+            "Received NS from UNSPEC, but message contains the \
+            Link-Layer Address option, dropping.",
+        )));
+    }
+
+    Ok((target_addr, has_ll_option))
+}
+
+// Return the destination IP and a Neighbor Advertisement, based on the data
+// provided in a Neighbor Solicitation. If we should not generate an NA in
+// response to the NS, then `None` is returned.
+//
+// See https://www.rfc-editor.org/rfc/rfc4861.html#section-7.2.4 for details on
+// the validation and construction performed here.
+fn construct_neighbor_advert<'a>(
+    na: &'a NeighborAdvertisement,
+    target_addr: &'a Ipv6Addr,
+    has_ll_option: bool,
+    src_ip: &'a Ipv6Addr,
+) -> Option<(Ipv6Addr, NdiscRepr<'a>)> {
+    // Drop the packet if the target address is not actually our own address.
+    if target_addr != na.ip() {
+        return None;
+    }
+
+    // Set the ROUTER flag, if required.
+    let mut flags = NdiscNeighborFlags::empty();
+    flags.set(NdiscNeighborFlags::ROUTER, na.is_router);
+
+    // Even though this is NA is in response to an NS, if the source IP is
+    // UNSPEC, we must _not_ set the SOLICITED flag.
+    let src_is_unspec = src_ip == &Ipv6Addr::ANY_ADDR;
+    flags.set(NdiscNeighborFlags::SOLICITED, !src_is_unspec);
```

Comment:
So I'm reading RFC 2461 4.4, and my reading of the `S` flag is that you don't set it if the **advertisement** is multicast, I don't think it care about how the NS came in. In fact, if a node sends an NS, isn't that by definition a solicitation for the target address? That is, the NA is in response to an NS, it was solicited. Am I missing something?

Reply:
So, I'm basing all this off RFC 4861, which I think is the updated version of 2461. That said, I wrote the code from this paragraph:

> If the source of the solicitation is the unspecified address, the
   node MUST set the Solicited flag to zero and multicast the
   advertisement to the all-nodes address.  Otherwise, the node MUST set
   the Solicited flag to one and unicast the advertisement to the Source
   Address of the solicitation.

Which is paragraph 3 of this section: https://www.rfc-editor.org/rfc/rfc4861.html#section-7.2.4. I read that as requiring us to send an NA to all-nodes, if we got an NS from the unspecified address. Am I misinterpreting that?

Reply:
I think actually that's the same thing as what you said:

> you don't set it if the advertisement is multicast

That's because we only multicast the NA if we got the NS from the unspecified address. So checking if the source is unspec is equivalent to checking if the destination is multicast. Does that sound right?

Reply:
Oh right, duh. I believe you are correct. Honestly IPv6 makes all this very hard to think about. At this point if guests are resolving the gateway correctly I'm good to go. My eyes are bleeding reading these RFCs.